src/renderer/gl_engine/tvgGlTessellator.cpp

```diff
@@ -1670,10 +1670,20 @@ void Stroker::doStroke(const PathCommand *cmds, uint32_t cmd_count, const Point
                 pts++;
             } break;
             case PathCommand::LineTo: {
+                // check if the line is a point
+                if (mStrokeState.prevPt == *pts) {
```

Comment:
LineTo target to same point needs to do nothing, I think this logical should only trigged in handleStrokeCap
think about this:
```
path->moveTo(10, 10);
path->lineTo(100, 10);
path->lineTo(100, 10);
path->lineTo(10,20);
```

>BTW. The code for the stroke tessellation is a little bit mess :(

Reply:
Did you specifically mean the strokeCap method? That’s where I initially started working on this issue, but it turned out not to be the best place, as we don’t want to draw caps when dealing solely with move operations.

In your example code, a cap would unnecessarily appear along with a lineTo to the same point, which shouldn’t happen.

Reply:
@RuiwenTang 
What do you think about this solution?

```c++
void Stroker::doStroke(const PathCommand *cmds, uint32_t cmd_count, const Point *pts, uint32_t pts_count)
{
    mResGlPoints->reserve(pts_count * 4 + 16);
    mResIndices->reserve(pts_count * 3);

    auto allowZeroLengthCap = false;

    for (uint32_t i = 0; i < cmd_count; i++) {
        switch (cmds[i]) {
            case PathCommand::MoveTo: {
                if (mStrokeState.hasMove) {
                    strokeCap(allowZeroLengthCap);
                    mStrokeState.hasMove = false;
                }
                mStrokeState.hasMove = true;
                mStrokeState.firstPt = *pts;
                mStrokeState.firstPtDir = {0.0f, 0.0f};
                mStrokeState.prevPt = *pts;
                mStrokeState.prevPtDir = {0.0f, 0.0f};
                pts++;
                allowZeroLengthCap = false;
            } break;
            case PathCommand::LineTo: {
                if (mStrokeState.prevPt == *pts) allowZeroLengthCap = true;
                this->strokeLineTo(*pts);
                pts++;
            } break;
            case PathCommand::CubicTo: {
                if (mStrokeState.prevPt == pts[2]) allowZeroLengthCap = true;
                this->strokeCubicTo(pts[0], pts[1], pts[2]);
                pts += 3;
            } break;
            case PathCommand::Close: {
                this->strokeClose();

                mStrokeState.hasMove = false;
            } break;
            default:
                break;
        }
    }
    strokeCap(allowZeroLengthCap);
}
```

and

```c++
void Stroker::strokeCap(bool allowZeroLength)
{
    if (!allowZeroLength && mStrokeState.firstPt == mStrokeState.prevPt) return;
    if (mStrokeCap == StrokeCap::Butt) return;

    if (mStrokeCap == StrokeCap::Square) {
        if (allowZeroLength && mStrokeState.firstPt == mStrokeState.prevPt) strokeSquarePoint(mStrokeState.firstPt);
        else {
            strokeSquare(mStrokeState.firstPt, {-mStrokeState.firstPtDir.x, -mStrokeState.firstPtDir.y});
            strokeSquare(mStrokeState.prevPt, mStrokeState.prevPtDir);
        }
    } else if (mStrokeCap == StrokeCap::Round) {
        if (allowZeroLength && mStrokeState.firstPt == mStrokeState.prevPt) strokeRoundPoint(mStrokeState.firstPt);
        else {
            strokeRound(mStrokeState.firstPt, {-mStrokeState.firstPtDir.x, -mStrokeState.firstPtDir.y});
            strokeRound(mStrokeState.prevPt, mStrokeState.prevPtDir);
        }
    }
}
```

Reply:
If so, I think only check some thing like `allowZeroLengthCap ` is enough.
Maybe the following code can work perfectly.
```
void Stroker::doStroke(const PathCommand *cmds, uint32_t cmd_count, const Point *pts, uint32_t pts_count)
{
    mResGlPoints->reserve(pts_count * 4 + 16);
    mResIndices->reserve(pts_count * 3);

    auto hasLineTo = false;

    for (uint32_t i = 0; i < cmd_count; i++) {
        switch (cmds[i]) {
            case PathCommand::MoveTo: {
                if (hasLineTo) { // check this, so we can skip if path only contains move instruction
                    strokeCap();
                    hasLineTo = false;
                }
                mStrokeState.hasMove = true;
                mStrokeState.firstPt = *pts;
                mStrokeState.firstPtDir = {0.0f, 0.0f};
                mStrokeState.prevPt = *pts;
                mStrokeState.prevPtDir = {0.0f, 0.0f};
                pts++;
                hasLineTo = false;
            } break;
            case PathCommand::LineTo: {
                hasLineTo = true;
                this->strokeLineTo(*pts);
                pts++;
            } break;
            case PathCommand::CubicTo: {
                 hasLineTo = true;
                this->strokeCubicTo(pts[0], pts[1], pts[2]);
                pts += 3;
            } break;
            case PathCommand::Close: {
                this->strokeClose();

                hasLineTo = false;
            } break;
            default:
                break;
        }
    }
    if (hasLineTo) strokeCap();
}

// strokeCap can handle if the path is zero length.
void Stroker::strokeCap()
{
    if (mStrokeCap == StrokeCap::Butt) return;

    if (mStrokeCap == StrokeCap::Square) {
        if (mStrokeState.firstPt == mStrokeState.prevPt) strokeSquarePoint(mStrokeState.firstPt);
        else {
            strokeSquare(mStrokeState.firstPt, {-mStrokeState.firstPtDir.x, -mStrokeState.firstPtDir.y});
            strokeSquare(mStrokeState.prevPt, mStrokeState.prevPtDir);
        }
    } else if (mStrokeCap == StrokeCap::Round) {
        if (mStrokeState.firstPt == mStrokeState.prevPt) strokeRoundPoint(mStrokeState.firstPt);
        else {
            strokeRound(mStrokeState.firstPt, {-mStrokeState.firstPtDir.x, -mStrokeState.firstPtDir.y});
            strokeRound(mStrokeState.prevPt, mStrokeState.prevPtDir);
        }
    }
}

```

Reply:
@RuiwenTang This solution seems fine. Additionally, I noticed that in this case, the ```hasMove``` flag in the ```State``` structure is only updated but never read. Perhaps we should remove this value as well? Or do you have any specific plans for this flag?

Reply:
Ok, it can be removed.